### ISSUE TITLE
Auto-fix has... gone haywire

### ISSUE THREAD

**Issue by msorens** [semgrep/semgrep#1917]
**Describe the bug**
Auto-fix for the rule given below sometimes works, but sometimes... well, you have to see it to believe it. Read on...

Input
```
class ReportDataService {

   isEmpty(data) {
    if (data === undefined || data.length === 0 || (data.length === 1 && data[0] === 'passed')) {
      return true;
    }
    return false;
  }

  isZero(count: number): boolean {
    if (count === 0) {
      return true;
    }
    return false;
  }

  isAllZeros(data) {
    const array = Object.keys(data).map(itm => data[itm]);
    const result = remove(array, function (value) {
      return value !== 0;
    });
    return this.isEmpty(result);
  }
}
```

Rule
```
rules:
- id: simple-booleans
  patterns:
    - pattern-either: 
      - pattern: |
          if ($EXPR) { return true; } else { return false; }
      - pattern: |
          if (!$EXPR) { return false; } else { return true; }
      - pattern: |
          if ($EXPR) { return true; } return false;
      - pattern: |
          if (!$EXPR) { return false; } return true;
  message: |
    simple booleans
  severity: WARNING
  fix: |
    return $EXPR;
```
Output
```
class ReportDataService {

isEmpty(data) {
    return data === undefined || data.length === 0 || (data.length === 1 && data[0] === 'passed';
  }

  isZero(count: number): boolean {
    if (count === 0) {
      return true;
    }
    return count === 0;) {
    const array = Object.keys(data).map(itm => data[itm]);
    const result = remove(array, function (value) {
      return value !== 0;
    });
    return this.isEmpty(result);
  }
}
```
As is often the case, a picture shows the problem more dramatically.
 - Two occurrences of the rule flagged: ✅ 
 - First occurrence fixed.  ✅ 
 - Second occurrence... mangled. ❌ 

<img width="1001" alt="image" src="https://user-images.githubusercontent.com/6817500/97516906-6441a080-1951-11eb-94a8-5fdee3280c79.png">

**To Reproduce**
The code mangling does not occur on semgrep.dev, but there a different problem manifests: duplicating the semicolon: https://semgrep.dev/s/5WGD/
To reproduce the main problem outlined above you can use the input fragment above or, because the repo is open source, you can use my real file: 

https://github.com/chef/automate/blob/28fe04ce84882059aa549771d2e74ba225680e8a/components/automate-ui/src/app/pages/+compliance/shared/reporting/report-data.service.ts#L118-L138

**Expected behavior**
Autofix should work correctly.

**What is the priority of the bug to you?**
Is this a P0 (blocking your adoption of Semgrep or workflow), P1 (important to fix or quite annoying), P2 (regular bug that should get fixed)?
**P2**

**Environment**
Semgrep 0.28.0

**Comment by ievans** [comment]
Mostly fixed as of 0.77, but the parenthesis gets dropped in the first example, which I believe is a duplicate with https://github.com/returntocorp/semgrep/issues/2902

input:
```
  isEmpty(data) {
    if (data === undefined || data.length === 0 || (data.length === 1 && data[0] === 'passed')) {
      return true;
    }
    return false;
  }
````

output:
```
  isEmpty(data) {
    return data === undefined || data.length === 0 || (data.length === 1 && data[0] === 'passed';
  }
```

**Comment by nmote** [comment]
Fixed by #7686
